src/libsyntax/parse/mod.rs

```diff
@@ -475,9 +352,23 @@ crate fn lit_token(lit: token::Lit, suf: Option<Symbol>, diag: Option<(Span, &Ha
     use ast::LitKind;
 
     match lit {
-       token::Byte(i) => (true, Some(LitKind::Byte(byte_lit(&i.as_str()).0))),
-       token::Char(i) => (true, Some(LitKind::Char(char_lit(&i.as_str(), diag).0))),
-       token::Err(i) => (true, Some(LitKind::Err(i))),
+        token::Byte(i) => {
+            let lit_kind = match unescape_byte(&i.as_str()) {
+                Ok(c) => LitKind::Byte(c),
+                Err((_, EscapeError::MoreThanOneChar)) => LitKind::Err(i),
+                Err(_) => LitKind::Byte(0),
```

Comment:
Recovery with `LitKind::Byte(0)` or `LitKind::Char('\u{FFFD}')` predates `LitKind::Err` and can probably be replaced by it.
I have hard time imagining the situation where it could give better diagnostics.

Reply:
Same for byte/string literals below.
Does anything in the test suite change if all of those are replaced with `LitKind::Err`?

Reply:
Only formatting test fails, with a new error of "format argument must be a string literal".

I presume this is because error message [here](https://github.com/rust-lang/rust/blob/d15fc173818ccea6f8d97af13a30d454f9e4b35c/src/libsyntax/ext/base.rs#L998-L1005) is not squashed, despite the fact that we produced erroneous literal for string. Will look into fixing that. I don't really know where  we should turn Literal::Err into Expr::Err, will try to figure this out, but pointers are appreciated!

Reply:
ah, that was super-trivial to fix :-)

Reply:
resolved